### PR TITLE
[exporter/splunkhec] add worker_queue for sending data to splunk

### PR DESCRIPTION
**Pull request by hvaghani221**
**Description:** 
This is an early draft to isolate batching the HEC event and sending it to splunk. 

Current Implementation
image

Suggested Implementation
image

The suggested approach has some advantages, like:
- It will not have the overhead recreating sub log or sub metric in case of failure when sending events to splunk (unless the worker queue is full)
- All events will be marshalled and compressed only once (unless worker queue is full)
- It makes it easier to add support for [HEC Indexer Acknowledgement](https://docs.splunk.com/Documentation/Splunk/9.0.1/Data/AboutHECIDXAck) since batching and sending are running on separate go routines, ack check can be done asynchronously.

### CODE REVIEW

**Comment by harshit-splunk** [comment]
@atoulme @dmitryax kindly have a look. Client.go file is very large so I have done some refactoring and moved some parts to a different file.

I have implemented this for log data only. If the design looks good, I will continue with metric and trace as well.

File: exporter/splunkhecexporter/client_test.go [hunk 1]
```diff
@@ -81,6 +85,12 @@ func newTestClientWithPresetResponses(codes []int, bodies []string) (*http.Clien
 	}, &headers
 }
 
+// type mockWorkerQueue struct {
+// 	// queue full
+// }
```
Comment (atoulme) [inline exporter/splunkhecexporter/client_test.go:90]: can remove

File: exporter/splunkhecexporter/exporter_test.go [hunk 2]
```diff
@@ -309,101 +309,102 @@ func generateLargeLogsBatch() plog.Logs {
...
-			if tt.wantErr {
-				assert.Error(t, err)
-				return
-			}
-
-			assert.NoError(t, err)
-		})
-	}
-}
+// TODO: Need to provide non-nil mocked worker queue
```
Comment (atoulme) [inline exporter/splunkhecexporter/exporter_test.go:312]: I would really want this test back before we get this refactor in.

File: exporter/splunkhecexporter/worker_queue.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	for request := range l.queue {
+		if !l.config.RetrySettings.Enabled {
+			if err := worker.Send(context.Background(), request); err != nil {
+				l.logger.Error("Failed to send data to splunk", zap.Error(err))
+				continue
+			}
+		}
+
+		var err error
+		for i := 0; i < l.config.WorkerQueueConfig.Retries; i++ {
```
Comment (atoulme) [inline exporter/splunkhecexporter/worker_queue.go:74]: There should not be a retry mechanism in the exporter, I think retries are done at the pipeline level by the collector.

File: exporter/splunkhecexporter/factory.go [hunk 4]
```diff
@@ -87,6 +87,12 @@ func createDefaultConfig() component.Config {
 		},
 		HealthPath:            splunk.DefaultHealthPath,
 		HecHealthCheckEnabled: false,
+		WorkerQueueConfig: WorkerQueueConfig{
```
Comment (atoulme) [inline exporter/splunkhecexporter/factory.go:90]: this config change needs discussion and should be done in a separate PR so we can discuss its merits.

File: exporter/splunkhecexporter/config_test.go [hunk 5]
```diff
@@ -101,6 +101,12 @@ func TestLoadConfig(t *testing.T) {
 				},
 				HealthPath:            "/services/collector/health",
 				HecHealthCheckEnabled: false,
+				WorkerQueueConfig: WorkerQueueConfig{
```
Comment (atoulme) [inline exporter/splunkhecexporter/config_test.go:104]: this config change needs discussion and should be done in a separate PR so we can discuss its merits.

File: exporter/splunkhecexporter/config.go [hunk 6]
```diff
@@ -46,6 +47,13 @@ type OtelToHecFields struct {
 	SeverityNumber string `mapstructure:"severity_number"`
 }
 
+type WorkerQueueConfig struct {
```
Comment (atoulme) [inline exporter/splunkhecexporter/config.go:50]: This should be out of this PR for now so we can have a PR just for this, and I don't think this is something we want. There are ways to retry defined at the pipeline level which are more effective than having something HEC specific. See https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go as a starting point.

Comment (harshit-splunk) [reply exporter/splunkhecexporter/config.go:50]: Actually, I was referring to it before I started working on this PR. I am planning to add HEC ACK support, but based on the current implementation, the splunk_hec exporter is either preparing a batch or sending the batch to splunk. 

HEC Ack requires checking the status from the indexer asynchronously. It shouldn't block batching/sending next data. Also, what should be next step if HEC Ack returns non-ok status? 
There are 2 possibilities that I can think of:

1. send all data to splunk and wait until it receives ACK status for all the batches.
    - Here, it needs to keep track of each batch and its buffer state for creating subLog/subMetric/subTrace and combine all the failed batches
    - While it is waiting for ACK status, the worker is not doing any other tasks, so it wastes the time and delays the next pushData call
2. Segregate batching and sending (current implementation)
    - It requires to create an internal queue. But, since batching and sending are done in different go routines, what should it do when sending events fails?
      - It can retry batches from the internal queue only (so it doesn't need to report back to the exporter helper)
      - It can report back failed batches during the next push cycle (it increases complexity as it needs to create subLog/subMetric/subTrace, so it also requires to add plog data into queue)

**Comment by atoulme** [comment]
@harshit-splunk I was discussing the state of the exporter over the PR earlier on the flaky test with @dmitryax. The refactoring changes you introduce in this PR are really good. If you can decouple away the queue, they would be awesome to bring in imo.

**Comment by atoulme** [comment]
If you can't, as this refactor looks complex - maybe we can chip at it with small focused PRs, one problem at a time. I'm game to help.

**Comment by harshit-splunk** [comment]
@atoulme I'll remove the worker queue from this PR
